ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt

```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
 public val AuthenticationInterceptors: RouteScopedPlugin<RouteAuthenticationConfig> =
     createRouteScopedPlugin("AuthenticationInterceptors", ::RouteAuthenticationConfig) {
 
-        on(AuthenticationHook) { call ->
-            val authConfig = call.application.plugin(Authentication).config
+        val providers = pluginConfig.providers
+        val authConfig = application.plugin(Authentication).config
 
+        val requiredProviders = authConfig
+            .findProviders(providers) { it == AuthenticationStrategy.Required }
+        val notRequiredProviders = authConfig
+            .findProviders(providers) { it != AuthenticationStrategy.Required } - requiredProviders
+        val firstSuccessfulProviders = authConfig
+            .findProviders(providers) { it == AuthenticationStrategy.FirstSuccessful } - requiredProviders
+        val optionalProviders = authConfig
+            .findProviders(providers) { it == AuthenticationStrategy.Optional } -
+            requiredProviders - firstSuccessfulProviders
+
+        on(AuthenticationHook) { call ->
             val authenticationContext = AuthenticationContext.from(call)
-            if (authenticationContext.principal != null) return@on
-
-            val configurations = pluginConfig.configurations.map { configurationName ->
-                authConfig.providers[configurationName] ?: throw IllegalArgumentException(
-                    if (configurationName == null) {
-                        "Default authentication configuration was not found"
-                    } else {
-                        "Authentication configuration with the name $configurationName was not found"
-                    }
-                )
-            }
-            for (provider in configurations) {
+            if (authenticationContext.principal<Principal>() != null) return@on
+
+            var count = 0
+            for (provider in requiredProviders) {
                 if (provider.skipWhen.any { skipCondition -> skipCondition(call) }) continue
 
                 provider.onAuthenticate(authenticationContext)
+                count++
+                if (authenticationContext._principal.principals.size < count) {
```

Comment:
What if the different(some optional) principals are saved and required are not? Could you add a test?

Reply:
These block checks only required providers. There are tests for different combinations of strategies in this PR.